### ISSUE TITLE
Restrictions on ServiceWorkers and SharedWorkers in same-origin prerendering

### ISSUE THREAD

**Issue by mfalken** [WICG/nav-speculation#44]
I started to lean toward allowing Shared Workers and Service Workers in same-origin prerendering, because they can be considered part of storage access, which we are allowing. 

https://github.com/jeremyroman/alternate-loading-modes/commit/789b4c07a7c1423b5964cede7535046564c529da makes a good point though that Shared Workers/Service Workers will have access to APIs that we intend to defer during prerendering, even in the same-origin case. For example, `Notification.requestPermission()`.

I previously had not considered this before, e.g., at: https://docs.google.com/document/d/1zY15k_wFTik2EoxBf3_RT7YjYpFMDaeNspy15n0rtww/edit#heading=h.2phlgq67a266.

For these workers, we can't say they are "in prerendering" or not, because they can be associated with multiple frames/browsing contexts.

Of course, even without service workers, a prerendered page can postMessage to a non-prerendered page, and have it use a restricted API. But in this case the non-prerendered page can be identified as the source of the API, for example, if we want to add a UI indicator that it's using a powerful permission-based API, etc.

I still want prerendered to be able to work with pages that use service workers, or else we will lose a lot of coverage and sites have to choose between prerendering and service workers.

One idea I have is to not let prerendered pages register Service Workers: this can hang. However, if a Service Worker is already installed, it can run and control the prerendered page. As long as the Service Worker has any clients that are non-prerendering, it can use APIs as normal. Otherwise, if all its clients are prerendering, it is considered the same as a prerendered browsing context.

I think this mostly makes sense because Service Workers cannot have any UI indicator anyway, as they can't be attributed to a UI widget. In fact considering that, it may be fine to just always consider service workers to be non-prerendering, as long as we block prerendered pages from registering service workers.

WDYT? This is all also still assuming same-origin prerendering. In the cross-origin case things are more complex, because we can get into cases of a service worker having both cross-origin prerendered pages and same-origin prerendered pages as clients.

**Comment by domenic** [comment]
Note that `Notification.requestPermission()` is `Window`-only. I think the problematic API is `self.registration.showNotification()`.

> One idea I have is to not let prerendered pages register Service Workers: this can hang. However, if a Service Worker is already installed, it can run and control the prerendered page. As long as the Service Worker has any clients that are non-prerendering, it can use APIs as normal. Otherwise, if all its clients are prerendering, it is considered the same as a prerendered browsing context.

This model seems reasonable. However, is the part of it that hangs on registration necessary? (For same-origin.) It seems like as long as we're doing some sort of check about whether or not all the clients are prerendering, that by itself gives us the tools we need to disallow `self.registration.showNotification()` and other similar APIs.

Such a check seems potentially racy since it could be crossing process boundaries, but since prerendering -> not-prerendering is a one-time transition, I think any races would resolve in favor of suppressing or delaying the notification, which is fine.

**Comment by mfalken** [comment]
> Note that `Notification.requestPermission()` is `Window`-only. I think the problematic API is `self.registration.showNotification()`.
> 
> > One idea I have is to not let prerendered pages register Service Workers: this can hang. However, if a Service Worker is already installed, it can run and control the prerendered page. As long as the Service Worker has any clients that are non-prerendering, it can use APIs as normal. Otherwise, if all its clients are prerendering, it is considered the same as a prerendered browsing context.
> 
> This model seems reasonable. However, is the part of it that hangs on registration necessary? (For same-origin.) It seems like as long as we're doing some sort of check about whether or not all the clients are prerendering, that by itself gives us the tools we need to disallow `self.registration.showNotification()` and other similar APIs.

Ah yes, good point, if we have the check that all clients are prerendering, I think we don't need to hang registration. I had two mitigations in mind and used them both.

> 
> Such a check seems potentially racy since it could be crossing process boundaries, but since prerendering -> not-prerendering is a one-time transition, I think any races would resolve in favor of suppressing or delaying the notification, which is fine.

Sounds good to me, yes.

**Comment by noamr** [comment]
I don't think we should have these restrictions...
It would make sense if a SW could show notification only if it had an active window, but that's not the case.
We'd have to create a special state for a SW, something like "semi-initialized" where it would do everything a SW would do but not show push notifications until at least one of its owners became fully rendered. I find that state to be a bit odd - the SW would expect that it had shown notifications, but in reality it did not.

Note that this is meant for the edge case where a same-origin prerendered page registers a SW which its initiator page did not.

I suggest something else:
- Expose prerendering state to the `Client` API, and `matchAll` etc
- The SW code can decide based on that whether to show the notification, even if the page that registered the SW was not activated yet.
- If this doesn't satisfy the developer, they can avoid this edge case explicitly in `speculationrules`.

Altermatively:
- Discard the prerendering page if it tries to register a SW that wasn't previously registered (thus avoiding the edge case altogether for a start).

Would be good to know from usage reports how common this use case is (and if it's indeed an edge case). Maybe as a start we can discard the prerendering and log something to the usage reports.

**Comment by nyaxt** [comment]
> Would be good to know from usage reports how common this use case is (and if it's indeed an edge case). Maybe as a start we can discard the prerendering and log something to the usage reports.

This sounds good to start with. @nhiroki wdyt?

**Comment by noamr** [comment]
A PR for this: https://github.com/WICG/nav-speculation/pull/131
Delaying installation if all clients are prerendering.

**Comment by nhiroki** [comment]
#131 switched from discarding service worker registration to deferring it. That sounds reasonable.

> So basically by prerendering a page with a service worker (in a user-agent initiated prerender) we have to choose between unwarranted installation of code, or potentially crippling some of the functionality of the page. Not sure what the solution is - seems like SW and prerendering are two features that don't play nice out of the box.

If a page doesn't work without service workers, it should wait until the completion of service worker registration before starting something. Therefore, I wonder if deferring service worker registration shouldn't break the page, it may not fully enjoy the power of prerendering though.

**Comment by domenic** [comment]
I'm wondering what principle we are using to make the decision here. Delaying registration seems good from the general principle of starting conservative, and preferring delay to discard or fail. So maybe it is fine for now.

But longer-term, would it be OK to install a service worker on a prerendered page? Is it dangerous in some way?

- It's fine to run code in a prerendered page. That is the point of prerendering.
- It's fine to cause permanent origin-wide effects in a prerendered page. We allow all storage APIs.

**Comment by noamr** [comment]
That's a very good question! I would say that we should consider each feature by:
- Does it have origin-wise/global side-effects?
- How exotic is it? (e.g. WebHID is more exotic than IndexedDB)
- Is it open-ended in scope?

Both storage and service-workers would answer yes to the first answer, and service-workers are probably more exotic than storage, at least currently. But the important bit is that service workers are open-ended: by enablilng them in a prerendered context it's unclear exactly what's enabled. Notifications are a good example of that, but may grow in the future.

To enable service-workers that are registered from a prerendered context, those service-workers would have to be "special" in a way - as in, know that they belong to an origin that is fully prerendered at the moment, and apply restrictions until at least one of their clients gets activated.

So to summarize I think it's OK to enable a feature that has origin-wide effects if they are fully understood, contained in scope, and in relatively common use. Storage falls under this category and (newly registered) service workers currently don't - though if this becomes an important enough use-case we can revise it and do the extra work to specify "service workers registered from a prerendering client"

**Comment by domenic** [comment]
> But the important bit is that service workers are open-ended: by enablilng them in a prerendered context it's unclear exactly what's enabled. Notifications are a good example of that, but may grow in the future.

I don't think that's the right way of thinking about it. Enabling service workers doesn't mean enabling notifications. It just means that we need to consider service worker features, like notifications, in our list of things to disable.

It's a tempting shortcut to just shut down all service worker features, but some of them, like fetch handling, seem pretty useful.

> To enable service-workers that are registered from a prerendered context, those service-workers would have to be "special" in a way - as in, know that they belong to an origin that is fully prerendered at the moment

All service workers belong to the origin that registered them (cross-origin service workers are impossible), so I don't think that is an issue.

> So to summarize I think it's OK to enable a feature that has origin-wide effects if they are fully understood, contained in scope, and in relatively common use. Storage falls under this category and (newly registered) service workers currently don't

Which part of this predicate do you think service workers fail?

- They seem fully understood; there are specs for all of their behaviors.
- They are contained in scope; everything they do is enumerated in specs and they don't have the ability to do more than what is in those specs.
- They are in relatively common use; [14%](https://chromestatus.com/metrics/feature/timeline/popularity/990) of page views use service workers.

**Comment by noamr** [comment]
In general I think we're in agreement. The difference between storage & SW I present here is mainly in complexity (which is maybe better wording than "contained scope"). In terms of common use, if 80% of sites relied on service workers to function then we would of course have to deal with it now...  I agree that once the basic features of prerendering work for users we should slowly open up things, and probably dedicated workers would come before service workers :) 

> All service workers belong to the origin that registered them (cross-origin service workers are impossible), so I don't think that is an issue.

We would have to define that a whole origin is "prerendered" and that the SW is prerendered, and how the lifecycle of that works, while so far a document was prerendered. Same goes for SharedWorkers (which are currently also deferred, but they can't do things like send notifications, which makes them a bit simpler).

> Which part of this predicate do you think service workers fail?
> 
> * They seem fully understood; there are specs for all of their behaviors.
> * They are contained in scope; everything they do is enumerated in specs and they don't have the ability to do more than what is in those specs.

Scope of service workers is very broad compared to local storage & IndexedDB, and is more likely to expand in scope going forward. I think "simple" is perhaps a better word than "contained".

**Comment by noamr** [comment]
Also, thinking of a scenario where you might have a bunch of installed service workers because you typed some things in the URL bar and never committed... there's something more involved about this than about having some extra IndexedDB databases or localStorage entries, but maybe it's just a feeling rather than something I can reason about.

**Comment by domenic** [comment]
> there's something more involved about this than about having some extra IndexedDB databases or localStorage entries, but maybe it's just a feeling rather than something I can reason about.

IMO this is a symptom of having the wrong perspective.

Service workers themselves, as a base capability, are not very complex or involved or intrusive. Their only base-level capability is programmatic fetch interception. In a way, you could think of them as having equivalent power to overriding `window.fetch` (plus, all other fetch-triggering APIs... which there are a lot of, including HTML elements). That's something pages that can run JavaScript can already do, with sufficient care and willingness to write ugly code.

There are a lot of features _in_ service workers which are complex and involved, such as background sync or notifications. But the same is true for document contexts.

IMO then, the better approach is to treat running service workers like running JavaScript/performing fetches/accessing storage: totally OK. And then treat features-in-service-workers on a case-by-case basis, like we do for document contexts.

I thought at first that delaying service worker modifications, per #131, would just be a shortcut that allows us to block off all those potentially-complex APIs, and deal with this problem space later. But I guess that's not really true, since we're allowing service worker code that was previously installed to continue running. So we still need to add specific blocks on notifications/background sync/etc. inside such already-installed service workers. Right?

**Comment by noamr** [comment]
> Service workers themselves, as a base capability, are not very complex or involved or intrusive. Their only base-level capability is programmatic fetch interception. In a way, you could think of them as having equivalent power to overriding `window.fetch` (plus, all other fetch-triggering APIs... which there are a lot of, including HTML elements). That's something pages that can run JavaScript can already do, with sufficient care and willingness to write ugly code.

It's about intercepting `window.fetch` for the whole origin. That's very different from what you can do on the document level. And unlike storage, it's intercepting/implicit (with storage other documents in the origin would have to explicitly read something from storage).

> 
> There are a lot of features _in_ service workers which are complex and involved, such as background sync or notifications. But the same is true for document contexts.
> 
> IMO then, the better approach is to treat running service workers like running JavaScript/performing fetches/accessing storage: totally OK. And then treat features-in-service-workers on a case-by-case basis, like we do for document contexts.
> 
> I thought at first that delaying service worker modifications, per #131, would just be a shortcut that allows us to block off all those potentially-complex APIs, and deal with this problem space later. But I guess that's not really true, since we're allowing service worker code that was previously installed to continue running. So we still need to add specific blocks on notifications/background sync/etc. inside such already-installed service workers. Right?

Possibly. Need to look at those features again and reconsider my perspective.... 
But I think it was a good call to block creating new SWs for now the same way we block creating dedicated/shared workers, as a way to limit prerendered page resources and start conservative.

**Comment by nhiroki** [comment]
Status Update: ServiceWorker APIs (`register`, `unregister`, `getRegistration`, `postMessage`) are now deferred (see #131 and #140). The Chromium has already implemented deferral of `register` and `postMessage`, and will implement the others soon [[crbug](https://crbug.com/1305120)].

**Comment by domenic** [comment]
This is done and in the spec!

**Comment by domenic** [comment]
> Out of curiosity, is this part done as well?

I don't think it is; reopening!

**Comment by domenic** [comment]
In the interests of issue triage, moving the remaining part to https://github.com/WICG/nav-speculation/issues/354.
